crates/subspace-networking/src/node_runner.rs

```diff
@@ -163,18 +168,47 @@ impl NodeRunner {
             let connected_peers_id_set = connected_peers.iter().cloned().collect();
             let reserved_peers_id_set = self.reserved_peers.keys().cloned().collect::<HashSet<_>>();
 
-            let missing_reserved_peer_ids =
-                reserved_peers_id_set.difference(&connected_peers_id_set);
+            let missing_reserved_peer_ids = reserved_peers_id_set
+                .difference(&connected_peers_id_set)
+                .collect::<Vec<_>>();
 
-            for peer_id in missing_reserved_peer_ids {
-                if let Some(addr) = self.reserved_peers.get(peer_id) {
-                    self.dial_peer(*peer_id, addr.clone());
+            if !missing_reserved_peer_ids.is_empty() {
+                // Free connection slots for reserved peers if required.
+                let peers_to_disconnect_number = (connected_peers_id_set.len()
+                    + missing_reserved_peer_ids.len())
+                .saturating_sub(self.max_established_total as usize);
+
+                if peers_to_disconnect_number > 0 {
+                    let peers_to_disconnect = connected_peers
+                        .clone()
+                        .into_iter()
+                        .take(peers_to_disconnect_number as usize);
+
+                    for peer_id in peers_to_disconnect {
+                        let result = self.swarm.disconnect_peer_id(peer_id);
+
+                        trace!(
+                            %local_peer_id,
+                            %peer_id,
+                            ?result,
+                            "Disconnect peer to free slots for reserved peers."
+                        );
+                    }
+                }
+
+                // Establish missing connections to reserved peers.
+                for peer_id in missing_reserved_peer_ids {
+                    if let Some(addr) = self.reserved_peers.get(peer_id) {
+                        self.dial_peer(*peer_id, addr.clone());
+                    }
                 }
```

Comment:
Coming back to the discussion about `if`s for readability. I believe here you have 3 extra idents because of that :)

Reply:
Only `if !missing_reserved_peer_ids.is_empty() ` could be considered as a `code comment`. Other if's prevent extra cloning.